src/core/model.js

```diff
@@ -157,28 +160,42 @@ export default class Model extends Object3D {
     defines,
     moduleSettings,
     defaultUniforms,
-    program
+    program,
+    shaderCache
   }) {
-    // Assign default shaders if none are provided
-    if (!vs) {
-      vs = MODULAR_SHADERS.vs;
-    }
-    if (!fs) {
-      fs = MODULAR_SHADERS.fs;
-    }
 
-    // Assign default uniforms (if any default shaders are being used)
-    if (vs === MONOLITHIC_SHADERS.vs || fs === MONOLITHIC_SHADERS.fs) {
-      defaultUniforms = defaultUniforms || MONOLITHIC_SHADERS.defaultUniforms;
-    }
+    this.getModuleUniforms = x => {};
+
+    if (!program) {
+      // Assign default shaders if none are provided
+      if (!vs) {
+        vs = MODULAR_SHADERS.vs;
+      }
+      if (!fs) {
+        fs = MODULAR_SHADERS.fs;
+      }
+
+      // Assign default uniforms (if any default shaders are being used)
+      if (vs === MONOLITHIC_SHADERS.vs || fs === MONOLITHIC_SHADERS.fs) {
+        defaultUniforms = defaultUniforms || MONOLITHIC_SHADERS.defaultUniforms;
+      }
 
-    const assembleResult = assembleShaders(this.gl, {vs, fs, modules, defines});
-    ({vs, fs} = assembleResult);
-    const {getUniforms} = assembleResult;
+      const assembleResult = assembleShaders(this.gl, {vs, fs, modules, defines});
+      ({vs, fs} = assembleResult);
 
-    this.getModuleUniforms = getUniforms || (x => {});
+      // Retrive compiled shaders from cache if exist.
+      if (shaderCache) {
+        vs = shaderCache.getVertexShader(this.gl, vs);
```

Comment:
What happens if the shaders are already `Shader` instances (i.e. not strings). Will ShaderCache handle that case?  Should we add a test case?

Also, not sure if supporting both Strings and Shaders is documented behavior, might be worth double checking the docs and make sure things are correctly described, at the same time as making sure the `shaderCache` parameter is added to `Model`.

Reply:
ShaderCache only handles 'strings', and this is already documented in shader-cache section.

Reply:
> ShaderCache only handles 'strings', and this is already documented in shader-cache section.

Yes that sounds right. To be clear, then I was probably correct in assuming that this code won't work when `Shader`s are passed in right. That is why it would be good to have a test case for `Model` being instantiated with `Shader`s instead of strings.